SPR-15993: Revisit meaning of response.close() in HTTP client

bclozel opened the pull request:
Prior to this issue, SPR-15920 added this new `close()` method which was
supposed to be called to clean resources after response processing.

This commit changes the meaning of that method: calling `close()` will
close the underlying HTTP connection. This has to be called if the
response body is not consumed by the application, since at that point
the underlying connection might be in an inconsistent state if shared in
a connection pool.

Issue: SPR-15993

bclozel commented:
Something we discussed today - WDYT @poutsma @rstoyanchev @sdeleuze ?

rstoyanchev commented:
I think it would be a good idea to explicitly mention in all those places in the javadoc that closing the connection is not needed if you use any methods that read the response body and that doing so cancels the benefits of connection pooling.

Also please add the same details in `WebClient#exchange` so we're covered in all 3 places.

oleg-alexeyev reviewed (approved):
While working around SPR-16011 we did exactly same change: `response.bodyToFlux()` -> `response.body(BodyExtractors.toFlux())`, effectively removing `close()` call from the chain, and everything started to work OK.

bclozel committed:
Revisit meaning of response.close() in HTTP client

Prior to this issue, SPR-15920 added this new `close()` method which was
supposed to be called to clean resources after response processing.

This commit changes the meaning of that method: calling `close()` will
close the underlying HTTP connection. This has to be called if the
response body is not consumed by the application, since at that point
the underlying connection might be in an inconsistent state if shared in
a connection pool.

Issue: SPR-15993
